sympy/solvers/solveset.py

```diff
@@ -2434,60 +2446,96 @@ def linear_coeffs(eq, *syms, **_kw):
 
-    >>> linear_coeffs(x + y*(z*(x*3 + 2) + 3), x)
-    [3*y*z + 1, y*(2*z + 3)]
+        >>> linear_coeffs(x + y*(z*(3*x + 2) + 3), x)
+        [3*y*z + 1, y*(2*z + 3)]
 
     But if there are nonlinear or cross terms -- even if they would
     cancel after simplification -- an error is raised so the situation
     does not pass silently past the caller's attention:
 
-    >>> eq = 1/x*(x - 1) + 1/x
-    >>> linear_coeffs(eq.expand(), x)
-    [0, 1]
-    >>> linear_coeffs(eq, x)
-    Traceback (most recent call last):
-    ...
-    NonlinearError: nonlinear term encountered: 1/x
-
-    >>> linear_coeffs(x*(y + 1) - x*y, x, y)
-    Traceback (most recent call last):
-    ...
-    NonlinearError: nonlinear term encountered: x*(y + 1)
+        >>> eq = 1/x*(x - 1) + 1/x
+        >>> linear_coeffs(eq.expand(), x)
+        [0, 1]
+        >>> linear_coeffs(eq, x)
+        Traceback (most recent call last):
+        ...
+        NonlinearError: nonlinear term: 1/x
+
+        >>> linear_coeffs(x*(y + 1) + 3, x, y)
+        Traceback (most recent call last):
+        ...
+        NonlinearError: nonlinear term: x*(y + 1)
+
+    An error is not raised when dependent symbols are passed, however:
+
+        >>> linear_coeffs(x + 2*x**2, x, x**2)
+        [1, 2, 0]
+
+    To allow coefficients to contain symbol-dependent factors which are
+    not strictly one of the symbols, use keyword `strict=False`.
+    A NonlinearError will still be reported if a term contains two
+    literal symbols as factors:
+
+        >>> linear_coeffs(x*(y**2 + 1) + x*y, x, y, strict=False)
+        Traceback (most recent call last):
+        ...
+        NonlinearError: nonlinear term: x*y
+
+    But the error will no longer raise if the unexpanded expression
+    does not contain such a term:
+
+        >>> eq, v = x*(y**2 + 1) + 3, [x, y]
+        >>> linear_coeffs(eq, *v, strict=False)
+        [y**2 + 1, 0, 3]
+
+    This is likely most useful when dealing with functions and their
+    derivatives since a derivative "has" the function in its expression
+    but it might make sense to treat them as being independent.
+
+    Regardless of the setting of `strict`, it should always be
+    true that the result will reconstruct an expression that is
+    equivalent to the input expression.
+
+        >>> from math import prod
+        >>> r = _  # result from above
+        >>> eq.equals(r[-1] + sum([prod(i) for i in zip(v, r)]))
+        True
     """
-    d = defaultdict(list)
+    from builtins import dict as _dict
+    from sympy.polys.matrices.linsolve import _lin_eq2dict
     eq = _sympify(eq)
+    if isinstance(eq, Equality):
+        eq = eq.rewrite(Add, evaluate=False)
+    if len(syms) == 1 and iterable(syms[0]) and not isinstance(syms[0], Basic):
+        raise ValueError('expecting one or more expressions')
     symset = set(syms)
     if len(symset) != len(syms):
         raise ValueError('duplicate symbols given')
     has = set(iterfreeargs(eq)) & symset
```

Comment:
@oscarbenjamin , is this of any value or should the who symset be used below with `if not eq.has_free_arg(symset):` used here?